### PR TITLE
Fix bug of gpu simulator

### PR DESCRIPTION
**Pull request by beizhansl**
1. fix bug of gpu simulator
2. prevent executing classical circuit on py_simu and gpu simulator

### CODE REVIEW

File: quafu/simulators/simulator.py [hunk 1]
```diff
@@ -82,6 +84,8 @@ def simulate(
             count_dict, psi = simulate_circuit(qc, psi, shots)
             
     elif simulator == "py_simu":
+        if qc.executable_on_backend == False:
+                raise QuafuError("classical operation only support for `qfvm_qasm`")
```
Comment (Zhaoyilunnn) [inline quafu/simulators/simulator.py:88]: minor: seems to have too many indents

File: src/qfvm/circuit.hpp [hunk 2]
```diff
@@ -100,6 +101,18 @@ void Circuit::add_op(QuantumOperator &op){
     }
 }
 
+vector<QuantumOperator> Circuit::gates(){
```
Comment (Zhaoyilunnn) [inline src/qfvm/circuit.hpp:104]: where's this function called

File: quafu/simulators/simulator.py (hunk 1 above)
Comment (beizhansl) [reply quafu/simulators/simulator.py:88]: my bad

File: src/qfvm/circuit.hpp (hunk 2 above)
Comment (beizhansl) [reply src/qfvm/circuit.hpp:104]: The gpu simulator will call: https://github.com/ScQ-Cloud/pyquafu/blob/93309771a5b7465416ec20ef9dbdfd6b5ad2ce2f/src/qfvm_gpu/cuda_simulator.cuh#L26C39-L26C39

**Pull request closed without merging**